Docs: add upgrade notes for Hilla 2.0 (#2226)
vaadin/docs: articles/lit/guides/upgrading/index.adoc

Spring Boot 3 / Framework 6 use new versions of third-party dependencies: Hibernate 6, Hibernate Validator 8, servlet containers - Jetty 11, Tomcat 10.1 and many others.

Spring provides the https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Migration-Guide[Dedicated Migration Guide for Spring-boot 3.0] and https://github.com/spring-projects/spring-framework/wiki/Upgrading-to-Spring-Framework-6.x[Upgrading to Spring Framework 6.x Guide].

To browse the full list of changes, see the https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Release-Notes[Spring-boot 3.0 Release Notes] and the https://github.com/spring-projects/spring-framework/wiki/What%27s-New-in-Spring-Framework-6.x[What's New in Spring Framework 6.x] page.

Below is a general overview of the changes needed for Spring-based Vaadin applications:

* *Upgrade Spring to the Latest*
+
You need to upgrade Spring versions to the latest, including the starter parent dependency:
+
.pom.xml
[source,xml]
----
<parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-parent</artifactId>

>     <version>3.0.3</version>

Author: ```suggestion
    <version>3.0.2</version>
```

This affects nullability in generated TypeScript code.

Note that it may be more practical to use the link:https://hilla.dev/docs/lit/reference/type-nullability/#nonnullapi[@NonNullApi] annotation for declaring nullability on the package level.

== Breaking Changes in Vaadin Components

* Badges no longer shrink by default; this can be overridden globally with CSS `[theme~="badge"] { flex-shrink:1; }`, or for specific instances with `badge.getStyle().set("flex-shrink", "1");` or `layout.setFlexShrink(badge, 1)`;
* Buttons no longer shrink by default; this can be overridden globally with CSS `vaadin-button { flex-shrink:1; }`, or for specific instances with `btn.getStyle().set("flex-shrink", "1");` or `layout.setFlexShrink(btn, 1)`;
* `CheckboxGroup::setItemLabelGenerator` no longer clears the current selection; `CheckboxGroup.clear()` can be used to clear value separately
* `BlurNotifier<CustomField>` type signature was corrected to `BlurNotifier<CustomField<T>>`
* Date Picker now uses the ISO 8601 date format (`yyyy-mm-dd`) as fallback for unsupported locales, instead of `mm-dd-yyyy`

> * Number Field's default width now matches that of other text input components; the old default can be brought back with CSS `vaadin-number-field { width:8em; }`

Reviewer: Does Number Field look like another compound adjective here?

Reviewer: I suppose that this document comes from Flow. Maybe the docs folks don't want to check language twice. You could ask them to first check the Flow version and then copy it.

== Preparations

=== Setting up Node.js

Installing an up-to-date version of Node.js 18 is recommended before starting with Hilla 2 migration. For the optimal and smooth application development experience, make sure to have `node` executables in the `PATH` environment variable. Here are the link:https://nodejs.dev/en/learn/how-to-install-nodejs/[Node.js installation] instructions to follow.

=== Maven wrapper

Hilla 2 requires either a Maven wrapper script in the application, or the `mvn` executable in the environment for configuring the endpoints parser and generator from the settings declared `pom.xml`.

If your project has `mvnw` and `mvnw.cmd` Maven wrapper scripts, Hilla 2 uses it to install and run Maven. The link:https://maven.apache.org/wrapper/[Maven Wrapper] website offers download links and instructions for installing it in a project.

For installing Maven in your system environment, see the link:https://maven.apache.org/install.html[Installing Apache Maven] page.

=== Hilla dependency upgrade

Upgrade the Hilla version in `pom.xml` to the latest release:

.pom.xml
[source,xml]
----

> <hilla.version>2.0.0.rc1</hilla.version>

Reviewer: Just a reminder to update to final

Spring Boot 3 / Framework 6 use new versions of third-party dependencies: Hibernate 6, Hibernate Validator 8, servlet containers - Jetty 11, Tomcat 10.1 and many others.

Spring provides the https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Migration-Guide[Dedicated Migration Guide for Spring-boot 3.0] and https://github.com/spring-projects/spring-framework/wiki/Upgrading-to-Spring-Framework-6.x[Upgrading to Spring Framework 6.x Guide].

To browse the full list of changes, see the https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Release-Notes[Spring-boot 3.0 Release Notes] and the https://github.com/spring-projects/spring-framework/wiki/What%27s-New-in-Spring-Framework-6.x[What's New in Spring Framework 6.x] page.

Below is a general overview of the changes needed for Spring-based Vaadin applications:

* *Upgrade Spring to the Latest*
+
You need to upgrade Spring versions to the latest, including the starter parent dependency:
+
.pom.xml
[source,xml]
----
<parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-parent</artifactId>

>     <version>3.0.0</version>

Reviewer: I see no reason to suggest version 3.0.0.

>     <version>3.0.0</version>

Author: The latest is `3.0.3`, let me use it.

```suggestion
    <version>3.0.3</version>
```

        InMemoryUserDetailsManager manager = new InMemoryUserDetailsManager();
        manager.createUser(User.withUsername("user")
                .password(passwordEncoder.encode("userPass"))
                .roles("USER").build());
        manager.createUser(User.withUsername("admin")
                .password(passwordEncoder.encode("adminPass"))
                .roles("USER", "ADMIN").build());
        return manager;
    }
}
----
+
In this example, `AuthenticationManagerBuilder`, used in Spring-boot 2.x, is replaced by `UserDetailsService`. And `http.authorizeRequests().antMatchers()` are replaced by `http.authorizeHttpRequests().requestMatchers()`.

== Java Version

*Java 17* or later is required. Below is an example of how to use this version:

[.example]
--
[source,xml]
----
<source-info group="Maven"></source-info>
<properties>
    <java.version>17</java.version>
    <!-- OR: -->
    <maven.compiler.source>17</maven.compiler.source>
    <maven.compiler.target>17</maven.compiler.target>
</properties>
----
[source,groovy]
----
<source-info group="Groovy"></source-info>
plugins {
    id 'java'
}

java {
    sourceCompatibility = 17
    targetCompatibility = 17
}

> ----

Reviewer: Isn't that misleading, given that Gradle is not supported?

--

== Maven Plugins

=== Hilla Maven Plugin

Starting from Vaadin 24, `vaadin-maven-plugin` alone does not support generating Hilla endpoints. Replace it with `hilla-maven-plugin` to enable using Hilla endpoint:

.pom.xml
[source,xml]
----
<groupId>dev.hilla</groupId>
<artifactId>hilla-maven-plugin</artifactId>
<version>${hilla.version}</version>
----

You can use Maven goals of the former `vaadin-maven-plugin` the same way from the `hilla-maven-plugin`.

This primarily affects hybrid applications, which have both Hilla client-side and Flow server-side views, and were created with Vaadin / Fusion 23 or earlier.

=== Third-Party Plugin Versions

Make sure that the versions of Maven plugins your project defines explicitly, are compatible with Java 17.

As an example, `nexus-staging-maven-plugin` requires a minimal version 1.6.13.

== SLF4J 2.0

Hilla 2, Vaadin 24 and Spring Boot 3 use SLF4J library version 2.0, which has breaking changes compared to earlier versions. See the https://www.slf4j.org/news.html[SLF4J release notes] for more information.

== Changes in Hilla Endpoints

> The new behavior of null type annotations (`@dev.hilla.Nonnull` and such) for arrays in endpoints and entity classes is:

Reviewer: ```suggestion
The behavior of null type annotations (`@dev.hilla.Nonnull` and such) has been fixed according to the https://docs.oracle.com/javase/specs/jls/se17/html/jls-9.html#jls-9.7.4[Java Language Specification]. In particular, it means that for arrays in endpoints and entity classes is:
```

- `@dev.hilla.Nonnull byte[]` is a nullable array type which items are non-null.

> - `byte @dev.hilla.Nonnull []` is non-null array type which items are nullable.

Reviewer: I'd use `String` rather than `byte` here, to highlight what is `null` and what is not.

* The default top and bottom margins of the `H1`...`H6` HTML elements have been removed; this change can be reverted by applying the following CSS:
+
.styles.css
[source,css]
----
h1,h2,h3,h4,h5,h6 { margin-top: 1.25em; }
h1 { margin-bottom: 0.75em; }
h2, h3, h4 { margin-bottom: 0.5em; }
h5 { margin-bottom: 0.25em; }
----

=== Changes in Web Component APIs

The following changes affect the client-side APIs of Vaadin components:

* The label on `vaadin-checkbox` and `vaadin-radio-button` must be set using the `label` property, as the default slot has been removed
* `vaadin-confirm-dialog.cancel` and `.reject` properties renamed `.cancelButtonVisible` and `.rejectButtonVisible`
* `vaadin-number-field` property `has-controls` renamed `step-buttons-visible`
* Deprecated `@vaadin/vaadin-**` (e.g. `@vaadin/vaadin-grid`) npm packages have been removed; use the new `@vaadin/**` instead (e.g. `@vaadin/grid`)
* Deprecated `**Element` legacy class aliases (e.g. `GridElement`) have been removed; use the plain component classes instead (e.g. `Grid`)

> * Deprecated misspelled `vaadin-icons` removed: `buss`, `funcion`, `megafone`, `palete`, `trendind-down`

Reviewer: 😮

Author: What do you mean? I guess your reaction is related with removing the `**Element` web component classes, because it gets confusing with React components, right?

Reviewer: Sorry, I was just surprised by the misspelled names! That's why I had marked the conversation as resolved. When pointing to a single line, GitHub doesn't highlight it correctly.
